crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4640,10 +4650,264 @@ mod tests {
 
         // TODO: once we support `sys.version_info` branches,
         // we can set `--target-version=py311` in this test
-        // and the inferred type will just be `BaseExceptionGroup` --Alex
+        // and the inferred type will just be `Unbound | BaseExceptionGroup` --Alex
         //
-        // TODO more precise would be `ExceptionGroup[TypeError | AttributeError]` --Alex
-        assert_public_ty(&db, "src/a.py", "e", "Unknown | BaseExceptionGroup");
+        // TODO more precise would be `Unbound | ExceptionGroup[TypeError | AttributeError]` --Alex
+        assert_public_ty(
+            &db,
+            "src/a.py",
+            "e",
+            "Unbound | Unknown | BaseExceptionGroup",
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definition_in_try_and_except_no_else() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 1
+            except:
+                x = 2
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // Either the `try` block ran to completion (an implicit empty `else` block),
+        // or the `except` block ran; either way, `x` is defined:
+        assert_public_ty(&db, "src/a.py", "x", "Literal[1, 2]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definition_in_try_and_multiple_excepts_no_else() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 1
+            except NameError:
+                x = 2
+            except TypeError:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // `x` may be unbound if the `except TypeError` branch is taken;
+        // it is bound if the the `try` block runs to completion or the `except NameError` branch is taken:
+        assert_public_ty(&db, "src/a.py", "x", "Unbound | Literal[1, 2]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definitions_with_pre_existing_definitions() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            e = 1
+            f = 2
+
+            try:
+                x
+            except NameError
```

Comment:
Should there be a colon after `NameError`? Is this a typo or expected? Same for other test cases.

Reply:
thanks!!